### PR TITLE
adding static features to time series data set API

### PR DESCRIPTION
**Pull request by canerturkmen**
This is a request for comments / discussion on updating the time series API to handle static features (item metadata in forecasting). In this PR, we propose to do this via several important changes to the `TimeSeriesDataFrame`.

1. We remove the `split_by_item` API as generic item indices do not necessarily have to follow a total ordering. In other words, for example, if each item id represents a distinct customer, then there are no use cases to really slice "all users up to user A."
2. We introduce the `static_features` property and metadata, which allows users to specify item metadata along with the time series data encapsulated in a `TimeSeriesDataFrame`. Embedding static features into the object has the advantage of separating data from logic, as well as avoiding introducing new arguments to all `fit` and `predict` functions in the time series module just for handling static features. However, it poses some difficulty in ensuring consistency in the time series data frame's data during data manipulation, which is what most of this PR deals with. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: timeseries/src/autogluon/timeseries/dataset/ts_dataframe.py [hunk 1]
```diff
@@ -112,11 +156,11 @@ def freq(self):
         return freq
 
     def iter_items(self) -> Iterable[Any]:
-        return iter(self.index.levels[0])
+        return iter(self._item_index)
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/dataset/ts_dataframe.py:159]: This also fixes a bug that we currently have on `master`, where `iter_items` is broken for subsets of a `TimeSeriesDataFrame` selected via `loc` :+1:

File: timeseries/src/autogluon/timeseries/dataset/ts_dataframe.py [hunk 2]
```diff
@@ -89,11 +104,40 @@ def __init__(self, data: Any, *args, **kwargs):
...
+    def static_features(self):
+        return self._static_features
+
+    @static_features.setter
+    def static_features(self, value: Optional[pd.DataFrame]):
+        # if the current item index is not a multiindex, then we are dealing with a single
+        # item slice. this should only happen when the user explicitly requests only a
+        # single item or during `slice_by_timestep`. In this case we do not set static features
+        if not isinstance(self.index, pd.MultiIndex):
+            return
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/dataset/ts_dataframe.py:129]: Maybe `return None` here to be more explicit?

File: timeseries/tests/unittests/test_ts_dataset.py [hunk 3]
```diff
@@ -20,13 +27,36 @@
...
+        [item_ids, datetime_index], names=["item_id", "timestamp"]
+    )
+    if static_features is not None:
+        return TimeSeriesDataFrame(
+            pd.Series(target, name="target", index=multi_inds).to_frame(),
+            static_features=static_features,
+        )
+    return TimeSeriesDataFrame(
+        pd.Series(target, name="target", index=multi_inds).to_frame(),
+    )
```
Comment (shchur) [inline timeseries/tests/unittests/test_ts_dataset.py:41]: Wouldn't this be equivalent to the following code (no if condition)?
```python
def _build_ts_dataframe(item_ids, datetime_index, target, static_features=None):
    multi_inds = pd.MultiIndex.from_product(
        [item_ids, datetime_index], names=["item_id", "timestamp"]
    )
    return TimeSeriesDataFrame(
        pd.Series(target, name="target", index=multi_inds).to_frame(),
        static_features=static_features,
    )
```

File: timeseries/tests/unittests/test_ts_dataset.py [hunk 4]
```diff
@@ -251,7 +235,9 @@ def test_subsequence(start_timestamp, end_timestamp, item_ids, datetimes, target
         (["2020-01-01 00:00:00", "2020-01-01 01:00:00", "2020-01-01 02:00:00"], "H"),
     ],
 )
-def test_when_dataset_constructed_from_dataframe_without_freq_then_freq_is_inferred(timestamps, expected_freq):
```
Comment (shchur) [inline timeseries/tests/unittests/test_ts_dataset.py:254]: It seems that you have `black` running with a different value of `line-length` (not 119 as in `pyproject.toml`), which generates these redundant formatting changes

**Review by shchur — COMMENTED** [review]
This looks good to me overall (except some minor changes requested in the comments)

File: timeseries/src/autogluon/timeseries/dataset/ts_dataframe.py (hunk 2 above)
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/dataset/ts_dataframe.py:129]: I would think this is semantically a bit cleaner. See https://stackoverflow.com/questions/15300550/return-return-none-and-no-return-at-all 
So to my understanding is that we `return None` when we have an `Optional` return type and we're explicitly returning a `None` somebody else may do something with. And `return` just means stop evaluating the function. What do you think?

File: timeseries/tests/unittests/test_ts_dataset.py (hunk 4 above)
Comment (canerturkmen) [reply timeseries/tests/unittests/test_ts_dataset.py:254]: 👍 blacking this file again.

File: timeseries/src/autogluon/timeseries/dataset/ts_dataframe.py (hunk 2 above)
Comment (shchur) [reply timeseries/src/autogluon/timeseries/dataset/ts_dataframe.py:129]: Somehow my read of that StackOverflow answer was that `return None` would be more appropriate here, but this is a really minor nitpick from my side, so I guess either option is fine.

**Review by shchur — APPROVED** [review]

**Pull request merged**